File: android/app/src/main/java/com/leo/jevpilot/TextCands.kt
package com.leo.jevpilot

/**
 * Pull typeable text out of the user's prompt (port of jevpilot/textcands.py).
 * Jev cannot generate text, so anything typed must already be in the prompt.
 */
object TextCands {
    private val QUOTES = Regex("\"([^\"]+)\"|'([^']+)'|„([^“”\"]+)[“”\"]|«([^»]+)»")
    val URL = Regex("\\b((?:https?://)?(?:[a-z0-9-]+\\.)+[a-z]{2,}(?:/\\S*)?)", RegexOption.IGNORE_CASE)
    private const val TRIGGERS =
        "suche(?:\\s+nach)?|such(?:\\s+nach)?|nach|search(?:\\s+for)?|for|tippe|tipp|type|" +
        "schreib(?:e)?|write|gib\\s+ein|eingeben|enter|namens|called|titel|title|" +
        "nachricht|message|text"
    private val TRIGGER_RE = Regex("\\b(?:$TRIGGERS)\\b[:\\s]+(.+)", RegexOption.IGNORE_CASE)
    private val STOP_RE = Regex(
        "[,.;!?]|\\s(?:und|and|dann|then|auf|on|in|bei|at|mit|with|über|via)\\s", RegexOption.IGNORE_CASE)
    private val LEAD_RE = Regex("^(?:in|ins|into|im|to|an)\\s+\\S+\\s+", RegexOption.IGNORE_CASE)

    private fun clean(s: String) = s.trim().trim('"', '\'', '„', '“', '”', '«', '»', ':').trim()

    fun candidates(prompt: String, limit: Int = 10): List<String> {
        val out = mutableListOf<String>()
        fun add(raw: String) {
            val s = clean(raw)
            if (s.length in 1..300 && out.none { it.equals(s, ignoreCase = true) }) out.add(s)
        }
        QUOTES.findAll(prompt).forEach { m -> m.groupValues.drop(1).firstOrNull { it.isNotEmpty() }?.let(::add) }
        TRIGGER_RE.findAll(prompt).forEach { m ->
            val rest = m.groupValues[1]
            val cut = STOP_RE.find(rest)
            add(if (cut != null) rest.substring(0, cut.range.first) else rest)
            add(LEAD_RE.replace(rest, ""))
            add(rest)
        }
        URL.findAll(prompt).forEach { add(it.groupValues[1]) }
        for (part in prompt.split(Regex("[,;\\n]"))) {
            val words = part.trim().split(Regex("\\s+")).filter { it.isNotEmpty() }
            for (n in listOf(2, 3, 1)) if (words.size > n) add(words.takeLast(n).joinToString(" "))
            if (words.size <= 6) add(part)
        }
        return out.take(limit)
    }

    fun urls(prompt: String): List<String> = URL.findAll(prompt).map { it.groupValues[1] }.toList()
}

File: jevpilot/textcands.py
"""Pull typeable text out of the user's prompt.

Jev cannot generate text, so anything the agent types must already be in the
prompt. We offer Jev a short list of prompt fragments and let it pick one.
"""
import re

QUOTES = re.compile(r'"([^"]+)"|\'([^\']+)\'|„([^“”"]+)[“”"]|«([^»]+)»')
URL = re.compile(r"\b((?:https?://)?(?:[a-z0-9-]+\.)+[a-z]{2,}(?:/[^\s]*)?)", re.I)
TRIGGERS = (
    r"suche(?:\s+nach)?|such(?:\s+nach)?|nach|search(?:\s+for)?|for|tippe|tipp|type|"
    r"schreib(?:e)?|write|gib\s+ein|eingeben|enter|namens|called|titel|title|"
    r"nachricht|message|text"
)
TRIGGER_RE = re.compile(rf"\b(?:{TRIGGERS})\b[:\s]+(.+)", re.I)
STOP_RE = re.compile(
    r"[,.;!?]|\s(?:und|and|dann|then|auf|on|in|bei|at|mit|with|über|via)\s", re.I
)


def _clean(s: str) -> str:
    return s.strip().strip("\"'„“”«»:").strip()


def text_candidates(prompt: str, limit: int = 10) -> list[str]:
    out: list[str] = []

    def add(s: str):
        s = _clean(s)
        if 1 <= len(s) <= 300 and s.lower() not in (x.lower() for x in out):
            out.append(s)

    for m in QUOTES.finditer(prompt):
        add(next(g for g in m.groups() if g))
    for m in TRIGGER_RE.finditer(prompt):
        rest = m.group(1)
        cut = STOP_RE.search(rest)
        add(rest[: cut.start()] if cut else rest)
        # "schreib in notepad hallo welt" -> "hallo welt"
        add(re.sub(r"^(?:in|ins|into|im|to|an)\s+\S+\s+", "", rest, flags=re.I))
        add(rest)
    for m in URL.finditer(prompt):
        add(m.group(1))
    # Clause tails, e.g. "google wetter berlin" -> "wetter berlin", "berlin".
    for part in re.split(r"[,;\n]", prompt):
        words = part.split()
        for n in (2, 3, 1):
            if len(words) > n:
                add(" ".join(words[-n:]))
        if len(words) <= 6:
            add(part)
    return out[:limit]


def urls_in(prompt: str) -> list[str]:
    return [m.group(1) for m in URL.finditer(prompt)]
